solr/core/src/test/org/apache/solr/search/TestMainQueryCaching.java

```diff
@@ -0,0 +1,313 @@
+    assertMetricCounts(response, matchAllDocs, expectFilterCacheInsertCount, expectFullSortCount, expectSkipSortCount, matchAllDocs ? ALL_DOCS : MOST_DOCS);
+  }
+
+  private static void assertMetricCounts(String response, boolean matchAllDocs, int expectFilterCacheInsertCount,
+                                         int expectFullSortCount, int expectSkipSortCount, int expectNumFound) {
+    Map<?, ?> res = (Map<?, ?>) fromJSONString(response);
+    Map<?, ?> body = (Map<?, ?>) (res.get("response"));
+    SolrCore core = h.getCore();
+    assertEquals("Bad matchAllDocs insert count", (matchAllDocs ? 1 : 0), coreToMatchAllDocsInsertCount(core));
+    assertEquals("Bad filterCache insert count", expectFilterCacheInsertCount, coreToInserts(core));
+    assertEquals("Bad full sort count", expectFullSortCount, coreToSortCount(core, "full"));
+    assertEquals("Bad skip sort count", expectSkipSortCount, coreToSortCount(core, "skip"));
+    assertEquals("Should have exactly " + expectNumFound, expectNumFound, (long) (body.get("numFound"))); // sanity check
+  }
+
+  @Test
+  public void testConcurrentMatchAllDocsInitialization() throws Exception {
+    final int nThreads = 20;
+    final ExecutorService executor = ExecutorUtil.newMDCAwareFixedThreadPool(nThreads, new SolrNamedThreadFactory(getTestName()));
+    final Future<?>[] followup = new Future<?>[nThreads];
+    for (int i = 0; i < nThreads; i++) {
+      final int myI = i;
+      followup[i] = executor.submit(() -> {
+        try {
+          // NOTE: we use cursorMark=* here because it prevents consulting the queryResultCache, which can interfere
+          // with DocSet fetching (which is what we care about in this test).
+          String response = JQ(req("q", MATCH_ALL_DOCS_QUERY, "request_id", Integer.toString(myI), "cursorMark", "*", "sort", "id asc"));
+          Map<?, ?> res = (Map<?, ?>) fromJSONString(response);
+          Map<?, ?> body = (Map<?, ?>) (res.get("response"));
+          assertEquals("Should have exactly " + ALL_DOCS, ALL_DOCS, (long) (body.get("numFound"))); // sanity check
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      });
+    }
+    try {
+      for (Future<?> f : followup) {
+        f.get(); // to access exceptions/errors
+      }
+    } finally {
+      executor.shutdown();
+      assertTrue(executor.awaitTermination(5, TimeUnit.SECONDS)); // tasks should already have completed
+    }
+    final SolrCore core = h.getCore();
+    Map<String, Object> liveDocsCacheMetrics = coreToLiveDocsCacheMetrics(core);
+    long inserts = (long) liveDocsCacheMetrics.get("inserts"); // the one and only liveDocs computation
+    long hits = (long) liveDocsCacheMetrics.get("hits"); // hits during the initial phase
+    long asyncHits = (long) liveDocsCacheMetrics.get("asyncHits");
```

Comment:
use the constants in SolrCache instead of strings here?

Reply:
FWIW I prefer readability in tests (String literals are plainly obvious).  And Yonik once pointed out that avoiding such constants in tests helps ensure we don't change something without seeing the impact (e.g. a client may depend on it).

Reply:
Yea, I can go either way. My thought was that if the constant ever moves, or the metric path moves, we'd want tests to fail compilation until they are updated. But this makes sense too. I'm not committed to either way.

Reply:
Interesting; I'm really ambivalent :-| ... I can see good arguments both ways. I'm inclined to let inertia win and leave it as-is?